src/workerinfo.js

```diff
@@ -34,57 +35,65 @@ class WorkerInfo {
     }
   }
 
-  async seen(provisionerId, workerType) {
+  async seen(provisionerId, workerType, workerGroup, workerId) {
     const expired = entity => Date.now() - new Date(entity.expires) > 24 * 60 * 60 * 1000;
     const expires = taskcluster.fromNow('5 days');  // temporary hard coded expiration
 
+    const updateExpiration = (entry) => {
+      entry.modify(entity => {
+        if (expired(entity)) {
+          entity.expires = expires;
+        }
+      });
+    };
+
+    const createEntry = async (entity, entry) => {
+      try {
+        await entity.create(entry);
+      } catch (err) {
+        // EntityAlreadyExists means we raced with another create, so just let it win
+        if (!err || err.code !== 'EntityAlreadyExists') {
+          throw err;
+        }
+      }
+    };
+
     const provisionerSeen = async (provisionerId) => {
       await this.valueSeen(provisionerId, async () => {
         // perform an Azure upsert, trying the update first as it is more common
         const provisioner = await this.Provisioner.load({provisionerId}, true);
-        if (provisioner) {
-          await provisioner.modify(entity => {
-            if (expired(entity)) {
-              entity.expires = expires;
-            }
-          });
 
-          return;
+        if (provisioner) {
+          return updateExpiration(provisioner);
         }
 
-        try {
-          await this.Provisioner.create({provisionerId, expires});
-        } catch (err) {
-          // EntityAlreadyExists means we raced with another create, so just let it win
-          if (!err || err.code !== 'EntityAlreadyExists') {
-            throw err;
-          }
-        }
+        createEntry(this.Provisioner, {provisionerId, expires});
       });
     };
 
     const workerTypeSeen = async (provisionerId, workerType) => {
       await this.valueSeen(`${provisionerId}/${workerType}`, async () => {
         // perform an Azure upsert, trying the update first as it is more common
         const wType = await this.WorkerType.load({provisionerId, workerType}, true);
-        if (wType) {
-          await wType.modify(entity => {
-            if (expired(entity)) {
-              entity.expires = expires;
-            }
-          });
 
-          return;
+        if (wType) {
+          return updateExpiration(wType);
         }
 
-        try {
-          await this.WorkerType.create({provisionerId, workerType, expires});
-        } catch (err) {
-          // EntityAlreadyExists means we raced with another create, so just let it win
-          if (!err || err.code !== 'EntityAlreadyExists') {
-            throw err;
-          }
+        createEntry(this.WorkerType, {provisionerId, workerType, expires});
+      });
+    };
+
+    const workerSeen = async (provisionerId, workerType, workerGroup, workerId) => {
+      await this.valueSeen(`${workerGroup}/${workerId}`, async () => {
```

Comment:
@jonasfj workerTypeSeen is already nesting under `${provisionerId}/${workerType}`. Wouldn't that interfere with the proposed nesting? I am thinking we could do the nesting under `${provisionerId}/${workerType}/${workerGroup}/${workerId}`. What do you think?